pyccl/halos/hmfunc.py

```diff
@@ -0,0 +1,653 @@
+                parametrizations).
+
+        Returns:
+            float or array_like: f(sigma_M) function.
+        """
+        raise NotImplementedError("Use one of the non-default "
+                                  "MassFunction classes")
+
+
+class MassFuncPress74(MassFunc):
+    """ Implements mass function described in 1974ApJ...187..425P.
+    This parametrization is only valid for 'fof' masses.
+
+    Args:
+        cosmo (:obj:`Cosmology`): A Cosmology object.
+    """
+    name = 'Press74'
+
+    def __init__(self, cosmo):
+        hmd = MassDef('fof', 'matter')
+        super(MassFuncPress74, self).__init__(cosmo,
+                                              hmd)
+
+    def _setup(self, cosmo):
+        self.norm = np.sqrt(2/np.pi)
+
+    def _check_mdef(self, mdef):
+        if mdef.Delta != 'fof':
+            return True
+        return False
+
+    def _get_fsigma(self, cosmo, sigM, a, lnM):
+        delta_c = 1.68647
+
+        nu = delta_c/sigM
+        return self.norm * nu * np.exp(-0.5 * nu**2)
+
+
+class MassFuncSheth99(MassFunc):
+    """ Implements mass function described in arXiv:astro-ph/9901122
+    This parametrization is only valid for 'fof' masses.
+
+    Args:
+        cosmo (:obj:`Cosmology`): A Cosmology object.
+    """
+    name = 'Sheth99'
+
+    def __init__(self, cosmo):
+        hmd = MassDef('fof', 'matter')
+        super(MassFuncSheth99, self).__init__(cosmo,
+                                              hmd)
+
+    def _setup(self, cosmo):
+        self.A = 0.21615998645
+        self.p = 0.3
+        self.a = 0.707
+
+    def _check_mdef(self, mdef):
+        if mdef.Delta != 'fof':
+            return True
+        return False
+
+    def _get_fsigma(self, cosmo, sigM, a, lnM):
+        delta_c = 1.68647
+
+        nu = delta_c/sigM
+        return nu * self.A * (1. + (self.a * nu**2)**(-self.p)) * \
+            np.exp(-self.a * nu**2/2.)
+
+
+class MassFuncJenkins01(MassFunc):
+    """ Implements mass function described in astro-ph/0005260.
+    This parametrization is only valid for 'fof' masses.
+
+    Args:
+        cosmo (:obj:`Cosmology`): A Cosmology object.
+    """
+    name = 'Jenkins01'
+
+    def __init__(self, cosmo):
+        hmd = MassDef('fof', 'matter')
+        super(MassFuncJenkins01, self).__init__(cosmo,
+                                                hmd)
+
+    def _setup(self, cosmo):
+        self.A = 0.315
+        self.b = 0.61
+        self.q = 3.8
+
+    def _check_mdef(self, mdef):
+        if mdef.Delta != 'fof':
+            return True
+        return False
+
+    def _get_fsigma(self, cosmo, sigM, a, lnM):
+        return self.A * np.exp(-np.fabs(-np.log(sigM) + self.b)**self.q)
+
+
+class MassFuncTinker08(MassFunc):
+    """ Implements mass function described in arXiv:0803.2706.
+
+    Args:
+        cosmo (:obj:`Cosmology`): A Cosmology object.
+        mass_def (:obj:`MassDef`): a mass definition object.
```

Comment:
So one quirk in the API here is that some mass functions do not take a mass def. I can imagine this causing weird things for people downstream. Thoughts on this?

Maybe each should have a default and we allow a new one but error check it?

Reply:
I can add a mass_def argument for those that only have one allowed mass definition for homogeneity. They won't work unless you pass the correct mass definition (which will be the default value). Make sense?

Reply:
Yep. This keeps the interface uniform which is the goal. Errors are user problems.

Reply:
Yep. This keeps the interface uniform which is the goal. Errors are user problems.

Reply:
Done